dear_petition/petition/models.py

```diff
@@ -309,6 +335,29 @@ def automatic_delete_date(self):
             hours=settings.CIPRS_RECORD_LIFETIME_HOURS
         )
 
+@receiver(signals.post_delete, sender=Batch)
+def cleanup_batch_client_on_delete(sender, instance, **kwargs):
+    """ If the client has no relevant batches, delete the client """
+    try:
+        if instance.client and len(instance.client.batches.all()) == 0:
+            instance.client.delete()
+    except Contact.DoesNotExist:
+        pass
+
+@receiver(signals.pre_save, sender=Batch)
+def cleanup_batch_client_pre_save(sender, instance, **kwargs):
```

Comment:
Another way we could approach this is by using a history tracker to track when the field has changed. It would simplify the logic.

Reply:
We can spin this off as a separate issue though.